titus-server-master/src/main/java/com/netflix/titus/master/supervisor/service/DefaultSupervisorOperations.java

```diff
@@ -0,0 +1,145 @@
+import com.netflix.titus.master.supervisor.model.event.MasterInstanceUpdateEvent;
+import com.netflix.titus.master.supervisor.model.event.SupervisorEvent;
+import com.netflix.titus.runtime.endpoint.metadata.CallMetadata;
+import com.netflix.titus.runtime.endpoint.metadata.CallMetadataUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Observable;
+
+@Singleton
+public class DefaultSupervisorOperations implements SupervisorOperations {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultSupervisorOperations.class);
+
+    private static final long GRACEFUL_SHUTDOWN_TIMEOUT_MS = 10_000;
+
+    private final LifecycleManager lifecycleManager;
+    private final MasterMonitor masterMonitor;
+    private final TitusRuntime titusRuntime;
+
+    @Inject
+    public DefaultSupervisorOperations(LifecycleManager lifecycleManager,
+                                       MasterMonitor masterMonitor,
+                                       TitusRuntime titusRuntime) {
+        this.lifecycleManager = lifecycleManager;
+        this.masterMonitor = masterMonitor;
+        this.titusRuntime = titusRuntime;
+    }
+
+    @Override
+    public List<MasterInstance> getMasterInstances() {
+        return masterMonitor.observeMasters().take(1).toBlocking().first();
+    }
+
+    @Override
+    public Optional<MasterInstance> findMasterInstance(String instanceId) {
+        return getMasterInstances().stream()
+                .filter(i -> i.getInstanceId().equals(instanceId))
+                .findFirst();
+    }
+
+    @Override
+    public MasterInstance getMasterInstance(String instanceId) {
+        return findMasterInstance(instanceId).orElseThrow(() -> SupervisorServiceException.masterInstanceNotFound(instanceId));
+    }
+
+    @Override
+    public Optional<MasterInstance> findLeader() {
+        return getMasterInstances().stream().filter(i -> MasterState.isLeader(i.getStatus().getState())).findFirst();
+    }
+
+    @Override
+    public Observable<SupervisorEvent> events() {
+        return masterMonitor.observeMasters()
+                .compose(ObservableExt.mapWithState(Collections.emptyMap(), this::buildEventList))
+                .flatMap(Observable::from);
+    }
+
+    @Override
+    public void restartMasterInstance(CallMetadata callMetadata) {
+        logger.warn("System shutdown requested");
+        titusRuntime.getSystemLogService().submit(SystemLogEvent.newBuilder()
+                .withPriority(SystemLogEvent.Priority.Warn)
+                .withMessage("System shutdown requested")
+                .withComponent(COMPONENT)
+                .withCategory(SystemLogEvent.Category.Transient)
+                .withContext(CallMetadataUtils.asMap(callMetadata))
+                .build()
+        );
+
+        // Make both threads non-daemon so they do not get terminated too early.
+        Thread controllerThread = new Thread("ShutdownController") {
```

Comment:
minor: may want to run this in a bounded thread pool to prevent abuse, but since we are about to `System.exit()`, I guess it doesn't matter much.

Reply:
Yes, in this case I purposefully created two non-daemon threads to handle the shutdown sequence.